### PR TITLE
Add MATCH-LINE macro

### PR DESCRIPTION
**Pull request by kilimanjaro**
This is just a bit of tidying. The easy case in the parser is that it is consuming a single line, dispatching on the first token and consuming a fixed template of successors (some required, some optional). This adds a simple bit of syntax for this.

### CODE REVIEW

File: src/parser.lisp [hunk 1]
```diff
@@ -369,56 +369,98 @@ INPUT-STRING that triggered the condition."
        (quil-parse-error "Got an unexpected token of type ~S ~
                           when trying to parse a program." tok-type)))))
 
+(defun idempotent-first (obj)
```
Comment (stylewarning) [inline src/parser.lisp]: `alexandria:ensure-car`

File: src/parser.lisp [hunk 2]
```diff
@@ -369,56 +369,98 @@ INPUT-STRING that triggered the condition."
...
+    (let ((spec-name (first spec))
+          (spec-type (second spec)))
+      (when spec-type
+        `((unless
+              ,(if opt
+                   `(or (null ,spec-name)
+                        (eql ',spec-type (token-type ,spec-name)))
+                   `(eql ,spec-type (token-type ,spec-name)))
+            (quil-parse-error
+             (format nil "Expected ~S token, but got ~~S." ,spec-type)
```
Comment (stylewarning) [inline src/parser.lisp]: compute this string at once if `spec-type` is known

File: src/parser.lisp [hunk 3]
```diff
@@ -369,56 +369,98 @@ INPUT-STRING that triggered the condition."
...
+        (first-line (gensym)))
+    (multiple-value-bind (required optional rest)
+        (alexandria:parse-ordinary-lambda-list token-specs
+                                               :allow-specializers t)
+      (let ((stripped-lambda-list (mapcar #'idempotent-first
+                                          token-specs))
+            (first-type (second (first required))))
+        `(let ((,all-lines ,lines))
+           ;; Check that we have a line to consume
+           (when (null ,all-lines)
```
Comment (stylewarning) [inline src/parser.lisp]: `endp`

File: src/parser.lisp [hunk 4]
```diff
@@ -369,56 +369,98 @@ INPUT-STRING that triggered the condition."
...
+             (quil-parse-error "Unexpectedly reached end of program~@[ (expected ~A instruction)]." ,first-type))
+           (let ((,first-line (first ,all-lines)))
+             ;; Check that we have the right number of tokens
+             (when (not (<= ,(length required)
```
Comment (stylewarning) [inline src/parser.lisp]: unless

File: src/parser.lisp [hunk 5]
```diff
@@ -369,56 +369,98 @@ INPUT-STRING that triggered the condition."
...
+                            (length ,first-line)
+                            ,@(unless rest
+                                (list (+ (length required) (length optional))))))
+               (quil-parse-error "Invalid line format~@[ for ~A instruction]." ,first-type))
```
Comment (stylewarning) [inline src/parser.lisp]: is there a better error message

Comment (kilimanjaro) [reply src/parser.lisp]: I gave a bit more context.

**Comment by kilimanjaro** [comment]
Also, props to @notmgsk for his "thought leadership".

File: src/parser.lisp [hunk 6]
```diff
@@ -661,145 +709,70 @@ INPUT-STRING that triggered the condition."
...
-      ;; Check that we are including a string (though we don't check if it's a valid path yet).
-      (unless (eql ':STRING (token-type path-tok))
-        (quil-parse-error "Expecting a pathname string to INCLUDE"))
-
-      ;; Return the object.
-      (values
-       (make-instance 'include :pathname (token-payload path-tok))
-       (rest tok-lines)))))
+  (match-line ((op :INCLUDE) (path :STRING)) tok-lines
+    (make-instance 'include :pathname (token-payload path))))
```
Comment (appleby) [inline src/parser.lisp]: Lovely

File: src/parser.lisp [hunk 7]
```diff
@@ -369,56 +369,104 @@ INPUT-STRING that triggered the condition."
...
+                              `(or (null ,spec-name)
+                                   (eql ',spec-type (token-type ,spec-name)))
+                              `(eql ,spec-type (token-type ,spec-name)))
+                       (quil-parse-error
+                        ,(if (keywordp spec-type)  ; If possible, format at expansion time.
+                             (format nil "Expected ~S token, but got ~~S." spec-type)
+                             `(format nil "Expected ~S token, but got ~~S." ,spec-type))
+                        (token-type ,spec-name)))))))))
+    (multiple-value-bind (required optional rest)
+        (alexandria:parse-ordinary-lambda-list token-specs
```
[the remark this answers is not in the document]
Comment (kilimanjaro) [reply src/parser.lisp]: A+ idea

**Pull request merged**